hw/mcu/nxp/kinetis/src/hal_timer.c

```diff
@@ -0,0 +1,428 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include <stdlib.h>
+#include <assert.h>
+
+#include "os/mynewt.h"
+#include "hal/hal_timer.h"
+#include "mcu/cmsis_nvic.h"
+
+#include "fsl_lptmr.h"
+
+#define KINETIS_TIMERS_MAX 2
+
+#if (FSL_FEATURE_LPTMR_CNR_WIDTH_IS_32B != 0)
+#error "The Kinetis hal_timer driver currently only supports 16-bit timers"
+#endif
+
+#if MYNEWT_VAL(TIMER_1) && (KINETIS_TIMERS_MAX < 2)
+#error "This MCU does not support TIMER_1"
+#endif
+
+/*
+ * This value for MCGIRCLK matches the current configuration for Kinetis BSPs
+ * where the slow clock is used and the divider is set to 1 (FCRDIV==0).
+ */
+#define MCGIRCLK_HZ 32768
+
+struct kinetis_hal_tmr {
+    LPTMR_Type *base;
+    uint16_t overflow; /* MSB of the current counter is handled manually */
+    TAILQ_HEAD(hal_timer_qhead, hal_timer) hal_timer_q;
+};
+
+#if MYNEWT_VAL(TIMER_0)
+struct kinetis_hal_tmr kinetis_tmr0 = {
+    .base = LPTMR0,
+};
+#endif
+#if MYNEWT_VAL(TIMER_1)
+struct kinetis_hal_tmr kinetis_tmr1 = {
+    .base = LPTMR1,
+};
+#endif
+
+static IRQn_Type kinetis_timer_irqs[] = LPTMR_IRQS;
+
+static struct kinetis_hal_tmr *kinetis_timers[KINETIS_TIMERS_MAX] = {
+#if MYNEWT_VAL(TIMER_0)
+    &kinetis_tmr0,
+#else
+    NULL,
+#endif
+#if MYNEWT_VAL(TIMER_1)
+    &kinetis_tmr1,
+#else
+    NULL,
+#endif
+};
+
+static uint32_t
+kinetis_tmr_read(struct kinetis_hal_tmr *tmr)
+{
+    return ((uint32_t)(tmr->overflow << 16) + LPTMR_GetCurrentTimerCount(tmr->base));
+}
+
+static void
+kinetis_tmr_set_period(struct kinetis_hal_tmr *tmr, uint32_t tick)
+{
+    /* XXX only set the lower 16-bits */
+    LPTMR_SetTimerPeriod(tmr->base, tick);
```

Comment:
Function LPTMR_SetTimerPeriod() should not be called when tick is 0 due to assert inside.
If I'm understanding it correctly with 32768 clock it can happen every 31 hours but I may be wrong.

Reply:
Yeah, good catch, it's interesting the CMR (compare) register cannot be zero, and also there is no control on the value o CNR (the counter), so I just checked the lower 16 bits for zero and incremented. Not sure how you came up with the 31 hours though, to it would seem more like 32768/3600 = 9.1 hours.

Reply:
31 hours was a mistake it's more like 36,4 hours (expiry is 32bits long) overflows at 4,294,967,296 ticks assuming 32768 ticks per second it makes 4,294,967,296 / 32768 (ticks in second) / 3600(seconds in hour) = 36,4 hours.

Anyway knowing how LPTMR_SetTimerPeriod() asserts when tick == 0 it could be that writing 0 to CMR register will not work (is not clear to me that CMR can't have 0 value) if that's the case I accept your solution that sometimes tick will be delayed.
But maybe writing 0 is OK and interrupt will happen and assert is for sanity only.
In that case knowing how it works and that 16 bits are used inside the function, something like this could be considered.
```c
LPTMR_SetTimerPeriod(tmr->base, (tick & 0xffff) | 0x10000);
```

Reply:
The datasheet says writing zero does not enable to timer and keeps CTF (the interrupt flag) on, so I think the way they thought this out is that it should always start counting from zero up to CMR.